File: android/app/src/main/java/org/mozilla/magnet/magnetapi/ApiAnalytics.java
package org.mozilla.magnet.magnetapi;

import android.content.Context;
import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import org.mozilla.magnet.BuildConfig;
import org.mozilla.magnet.api.Api;
import org.mozilla.magnet.tracking.*;

import java.util.Map;

public class ApiAnalytics extends Api {
    private static final String TAG = "ApiAnalytics";
    private final Context mContext;

    private final Api mApiPreferences;
    private final Analytics mGa;

    ApiAnalytics(Context aContext) {
        super(aContext);
        mContext = aContext;
        mGa = new Analytics(aContext, BuildConfig.GA_TRACKER_ID);
        mApiPreferences = new ApiPreferences(aContext);
    }

    @Override
    public void post(String aPath, final Object aData, final Callback aCallback) {
        mApiPreferences.get("", new Callback() {
            @Override
            public void resolve(Object aPrefData) {
                if (!(aPrefData instanceof JSONObject)) {
                    aCallback.reject("expected JSONObject from preferences API");
                    return;
                }

                JSONObject prefData = (JSONObject)aPrefData;
                boolean enableTelemetry = prefData.optBoolean("enableTelemetry");

                if (enableTelemetry) {

                    // Given the data, translate it to a 'CallableTracker'
                    // instance. Will throw if the data can not be converted
                    // into a tracker object.
                    asCallableTracker(aData).call(mGa);
                }

                aCallback.resolve(true);
            }

            @Override
            public void reject(String aError) {
                Log.d(TAG, "pref API call failed" + aError);
                aCallback.reject(aError);
            }
        });
    }

    public static TrackEvent createEvent(String aCategory, String aAction) {
      return new TrackEvent(aCategory, aAction);
    }

    public static CallableTracker asCallableTracker(Object aObject) {
        if (aObject instanceof CallableTracker) {
            return (CallableTracker)aObject;
        }

        if (!(aObject instanceof Map)) {
            throw new IllegalArgumentException("API request must be a 'Map' or instance of 'CallableTracker'");
        }

        Map<String, Object> data = (Map<String, Object>)aObject;

        String trackingType = (String)data.get("type");

        if (trackingType == null) {
            throw new IllegalArgumentException("Unspecified 'type' paramter");
        }

        switch(trackingType) {
            case "event":
                return TrackEvent.fromMap(data);
            case "timing":
                return TrackTiming.fromMap(data);
            case "screenview":
                return TrackScreenView.fromMap(data);
            default:
                throw new IllegalArgumentException("Unknown 'type' " + trackingType);
        }
    }
}

File: ios/ApiAnalytics.swift
//
//  ApiAnalytics.swift
//  Magnet
//
//  Created by sam on 11/21/16.
//  Copyright © 2016 Mozilla. All rights reserved.
//

import Foundation
import SwiftyJSON

class ApiAnalytics: ApiBase {
  let preferencesApi = ApiPreferences();
  let analytics = Analytics();
  override init() {
    super.init()
  }
  
  override func post(path: String, data: NSDictionary, callback: ApiCallback) {
    func onPreferencesReceived(json: JSON) {
      debugPrint("RECEIVED PREFERENCES")
      
      // If telemetry is not enabled, return early.
      if !asBoolean(json["enableTelemetry"].string) {
        callback.onSuccess(true);
        return
      }
      
      let request = JSON(data);
      
      if let type = request["type"].string {
        switch(type) {
        case "timing":
          TrackTiming.fromJSON(request).call(analytics)
        case "event":
          TrackEvent.fromJSON(request).call(analytics)
        case "screenview":
          TrackScreenView.fromJSON(request).call(analytics)
        default:
          
          callback.onError("Unkown tracking type: '" + type + "'")
          return
        }
        
        callback.onSuccess(true);
      } else {
        
        callback.onError("Unspecified tracking 'type' parameter")
      }
    }
    
    preferencesApi.get("", callback:
      ApiCallback(success: onPreferencesReceived, error: callback.onError));
  }
}


func asBoolean(string: String?) -> Bool {
  if let _string = string {
    switch(_string) {
    case "0": return false
    case "1": return true
    case "true": return true
    case "false": return false
    default: return false
    }
  }
  
  return false
}
